library/blob.c

```diff
@@ -2028,99 +2026,128 @@ static int eblob_write_prepare_disk(struct eblob_backend *b, struct eblob_key *k
 			copy, copy_offset, old);
 
 err_out_exit:
+	return err;
+}
+
+
+/**
+ * eblob_write_prepare_disk() - allocates space for new record
+ * It allocates new bases, commits headers and manages overwrites/appends.
+ * NB! Function takes "backend" lock.
+ */
+static int eblob_write_prepare_disk(struct eblob_backend *b, struct eblob_key *key,
+		struct eblob_write_control *wc, uint64_t prepare_disk_size,
+		enum eblob_copy_flavour copy, uint64_t copy_offset, struct eblob_ram_control *old,
+		size_t defrag_generation)
+{
+	pthread_mutex_lock(&b->lock);
+	int err = eblob_write_prepare_disk_nolock(b, key, wc, prepare_disk_size,
+	                                          copy, copy_offset, old, defrag_generation);
 	pthread_mutex_unlock(&b->lock);
-	eblob_dump_wc(b, key, wc, "eblob_write_prepare_disk", err);
 	return err;
 }
 
+
+/**
+ * write_prepare_inplace() - prepare space in existing base. Non-blocking.
+ */
+static int write_prepare_inplace(struct eblob_backend *b, struct eblob_key *key,
+		uint64_t flags, struct eblob_write_control *wc) {
+	// We've got a key which will be overwritten, make sure it has valid flags.
+	// We overwrite flags to what user has provided dropping all existing on-disk flags since
+	// given key will be fully overwritten and we do not care about its old content anymore.
+	// The same logic is performed in @eblob_write_prepare_disk(), but it also allocates new space.
+
+	int err = 0;
+	uint64_t new_flags = 0;
+	new_flags = eblob_validate_ctl_flags(b, flags);
+	new_flags |= BLOB_DISK_CTL_UNCOMMITTED;
+
+	if (wc->flags == new_flags) {
+		goto out_stat;
+	}
+
+	if (!(wc->flags & BLOB_DISK_CTL_UNCOMMITTED)) {
+		// update stats if the found key is committed, because now it becomes uncommitted.
+		eblob_stat_inc(wc->bctl->stat, EBLOB_LST_RECORDS_UNCOMMITTED);
+		eblob_stat_add(wc->bctl->stat, EBLOB_LST_UNCOMMITTED_SIZE,
+		               wc->total_size + sizeof(struct eblob_disk_control));
+
+		eblob_stat_inc(b->stat_summary, EBLOB_LST_RECORDS_UNCOMMITTED);
+		eblob_stat_add(b->stat_summary, EBLOB_LST_UNCOMMITTED_SIZE,
+		               wc->total_size + sizeof(struct eblob_disk_control));
+	}
+
+	wc->flags = new_flags;
+	err = eblob_commit_disk(b, key, wc, 0);
+	if (err)
+		goto out_exit;
+
+	err = eblob_commit_ram(b, key, wc);
+	if (err)
+		goto out_exit;
+
+out_stat:
+	eblob_stat_inc(b->stat, EBLOB_GST_PREPARE_REUSED);
+out_exit:
+	return err;
+}
+
+
 /**
  * eblob_write_prepare() - prepare phase reserves space in blob file.
  */
 int eblob_write_prepare(struct eblob_backend *b, struct eblob_key *key,
 		uint64_t size, uint64_t flags)
 {
-	FORMATTED(HANDY_TIMER_SCOPE, ("eblob.%u.disk.write.prepare", b->cfg.stat_id));
 	struct eblob_write_control wc = { .offset = 0 };
 	struct eblob_ram_control old;
 	int err;
 	size_t defrag_generation = 0;
 
-	EBLOB_WARNX(b->cfg.log, EBLOB_LOG_DEBUG,
```

Comment:
Why was it moved?

Reply:
we dereference b and after that check that it's not NULL. stupid.

Reply:
Could you please move this change into a separate commit?

Reply:
returned